MiniTwitch.Common/AsyncEventCoordinator.cs

```diff
@@ -0,0 +1,138 @@
+﻿namespace MiniTwitch.Common;
+
+/// <summary>
+/// Provides mechanisms to asynchronously wait for an event to occur
+/// </summary>
+/// <typeparam name="TEnum">The enum type to operate on</typeparam>
+public class AsyncEventCoordinator<TEnum> : IDisposable
+    where TEnum : struct
+{
+    private readonly Dictionary<int, SemaphoreSlim> _locks;
+
+    /// <summary>
+    /// Initializes a new instance of AsyncEventCoordinator
+    /// </summary>
+    /// <param name="initialCount">The initial semaphore request count</param>
+    /// <exception cref="NotSupportedException"></exception>
+    public AsyncEventCoordinator(int initialCount = 0)
+    {
+        if (!typeof(TEnum).IsEnum)
+            throw new NotSupportedException($"The type {typeof(TEnum)} is not an enum");
+
+        TEnum[] values = (TEnum[])Enum.GetValues(typeof(TEnum));
+        _locks = new(values.Length);
+        for (int i = 0; i < values.Length; i++)
+        {
+            _locks[(int)(object)values[i]] = new SemaphoreSlim(initialCount);
+        }
+    }
+
+    /// <summary>
+    /// Asynchronously wait for the <see cref="TEnum"/> value to be released
+    /// </summary>
+    /// <param name="value">The value to wait on</param>
+    /// <param name="timeout">The wait timeout, after which <see langword="false"/> is returned</param>
+    /// <param name="cancellationToken">The cancellation token</param>
+    /// <returns><see langword="true"/> if the wait was successful, otherwise <see langword="false"/></returns>
+    public Task<bool> WaitFor(TEnum value, TimeSpan timeout, CancellationToken cancellationToken = default)
+    {
+        int key = (int)(object)value;
+        return _locks[key].WaitAsync(timeout, cancellationToken);
+    }
+
+    /// <summary>
+    /// Asynchronously wait for any of the <see cref="TEnum"/> values to be released
+    /// </summary>
+    /// <param name="values">The values to wait on</param>
+    /// <param name="timeout">The wait timeout, after which <see langword="default"/> is returned</param>
+    /// <param name="cancellationToken">The cancellation token</param>
+    /// <returns>The <see cref="TEnum"/> value that was released, or <see langword="default"/> if the task timed out</returns>
+    public async Task<TEnum> WaitForAny(IEnumerable<TEnum> values, TimeSpan timeout, CancellationToken cancellationToken = default)
+    {
+        TEnum[] valuesArr = values.ToArray();
+        Task<bool>[] tasks = valuesArr.Select(x => WaitFor(x, timeout, cancellationToken)).ToArray();
+        Task<bool> completed;
+        try
+        {
+            completed = await Task.WhenAny(tasks).WaitAsync(timeout, cancellationToken);
```

Comment:
Does this compile? What is `.WaitAsync` on a bool?

Reply:
It does compile. WaitAsync here returns the task that completed first, that's what the ReferenceEquals() below is for

Reply:
I don't think you need `WaitAsync` here, you already pass `timeout` and `cancellationToken` into `WaitFor`

Reply:
`WaitFor` will return a result regardless, which is a "completed" state for `WhenAny`, returning it as the completed task whether it times out or not

Reply:
Wouldn't `WaitFor` throw `OperationCanceledException` in case cancellationToken is activated? I doubt it returns anything on timeout as well

Reply:
It doesn't throw an exception, but I found that the completed task is the only one that will pass the `tasks[i].IsCompleted && tasks[i].Result == true` condition. So you're right in saying I don't need `WaitAsync`. Thanks